packages/preferences/src/browser/preferences-widget.ts

```diff
@@ -0,0 +1,252 @@
+                name: group.name.toString(),
+                preferences: preferencesArray,
+                isExpanded: false
+            });
+        }
+
+        this.update();
+    }
+
+    protected onCloseRequest(msg: Message): void {
+        this.applicationShell
+            .getWidgets("right")
+            .forEach(widget => {
+                if (widget.id.endsWith('settings.json')) {
+                    widget.close();
+                }
+            });
+        this.applicationShell.collapsePanel("right");
+        super.onCloseRequest(msg);
+    }
+
+    protected render(): h.Child {
+        const preferenceGroups = [];
+        for (const preferenceGroup of this.preferencesGroups) {
+            preferenceGroups.push(this.renderPreferenceGroup(preferenceGroup));
+        }
+        return h.div({className: "preferences-container"}, ...preferenceGroups);
+    }
+
+    protected renderPreferenceGroup(group: PreferenceGroup): h.Child {
+        return h.div(
+            {className: "preference-group"},
+            h.div(
+                {
+                    className: "preference-group-header-div",
+                    onclick: () => {
+                        group.isExpanded = !group.isExpanded;
+                        this.update();
+                    }
+                },
+                group.name,
+                h.div({className: group.isExpanded ? "icon fa fa-caret-down" : "icon fa fa-caret-right"})
+            ),
+            group.isExpanded ? this.createPreferencesList(group.preferences) : ""
+        );
+    }
+
+    protected createPreferencesList(preferences: Preference[]): h.Child {
+        const preferenceItems: h.Child[] = [];
+        preferences.forEach(preference => {
+            preferenceItems.push(this.createPreferenceItem(preference));
+        });
+        return h.div({className: "preferences-list"}, ...preferenceItems);
+    }
+
+    protected createPreferenceItem(preference: Preference): h.Child {
+        const preferenceName: string = preference.name;
+        const nameSpan = h.span(
+            {
+                className: this.preferenceService.get(preferenceName) ? 'preference-item-name-saved' : 'preference-item-name-unsaved',
+                title: preference.property.description
+            },
+            preferenceName);
+        const property = preference.property;
+        const onBlur = function (event: FocusEvent): void {
+            const relatedTarget: any = event.relatedTarget;
+            if (relatedTarget && relatedTarget.className.startsWith('preference-item-value-')) {
+                return;
+            }
+            PreferencesWidget.hideElement(preferenceName);
+        };
+        let valueContainer: h.Child;
+        if (property.type === 'boolean' || property.enum) {
+            valueContainer = this.createEnumItems(preferenceName, property);
+        } else {
+            valueContainer = this.createInputValuePanel(preferenceName, property, onBlur);
```

Comment:
Please use the standard menu (`ContextMenuRenderer`) which will make sure it looks and feels the same as else where. Also it will make sure that it gets rendered properly (e.g. above the cursor when you are close to the bottom).
Like in VS Code it should propose the options when there is a finite set of alternatives (boolean, enums) otherwise copy the element over and select the value on the right.

Reply:
Is it possible to set different context menus to different tree items in the same tree?

Reply:
Yes, you can directly use the context menu renderer. I.e. without going through the menu registry.

Reply:
Setting non enum\boolean values directly to file will not give an ability to validate values. Some properties might not have default values, so in that case we wont't know about value type.

Reply:
That should be solved by enabling the JSON schema support.